src/shared/reboot-util.c

```diff
@@ -107,3 +108,43 @@ int shall_restore_state(void) {
 
         return r > 0 ? ret : true;
 }
+
+static int xen_kexec_loaded(void) {
+#if HAVE_XENCTRL
+        xc_interface *xch;
+        int r;
+
+        r = dlopen_xenctrl();
+        if (r < 0)
+                return r;
+
+        if (access("/proc/xen", F_OK) < 0)
+                return log_debug_errno(errno, "Cannot access /proc/xen: %m");
+
+        xch = sym_xc_interface_open(NULL, NULL, 0);
+        if (!xch)
+                return log_debug_errno(errno, "Failed to open Xen control interface: %m");
+
+        r = sym_xc_kexec_status(xch, KEXEC_TYPE_DEFAULT);
+
+        sym_xc_interface_close(xch);
+
+        return r;
+#else
+        return -EOPNOTSUPP;
+#endif
+}
+
+bool kexec_loaded(void) {
+       _cleanup_free_ char *s = NULL;
+       int r;
+
+       r = xen_kexec_loaded();
+       if (r >= 0)
+               return r != 0;
+
+       if (read_one_line_file("/sys/kernel/kexec_loaded", &s) < 0)
+               return false;
```

Comment:
i would very much prefer if we would put this check first, as it is the more common case.

Reply:
also, if you touch this, we nowadays prefer if unexpected errors are debug logged. i.e. ENOENT we kinda expect, any other errors we don't. hence, please change this to:

```c
r = read_one_line_file("/sys/kernel/kexec_loaded", &s);
if (r < 0) {
        if (r != -ENOENT) 
                log_debug_errno("Unable to read /sys/kernel/kexec_loaded, ignoring: %m");
        return false;
}
if (s[0] == '1'))
        return true;

return xen_kexec_loaded();
```

Reply:
> i would very much prefer if we would put this check first, as it is the more common case.

No: the non-existence of `/sys/kernel/kexec_loaded` doesn't say anything about whether the hypervisor has loaded a kexec kernel. And even if a user somehow loads a kernel by hand into linux (and thus `/sys/kernel/kexec_loaded` contain 1), kexec-ing with linux would just fail, it's really the hypervisor that needs to be contacted at all, when running under Xen.

Reply:
please make the log_debug_errno() change as mentioned, i.e. if you touch this, please adjust this to our current coding style.

(this is independent on the order in which you do the two questions)